### ISSUE TITLE
`vue/require-default-prop`: `defineModel` support added in 10.11.0 has no valid fix

### ISSUE THREAD

**Issue by hymair** [vuejs/eslint-plugin-vue#3135]
**Checklist**

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have read the [FAQ](https://eslint.vuejs.org/user-guide/#faq) and my problem is not listed.

**Tell us about your environment**

- **ESLint version:** 10.10.0
- **eslint-plugin-vue version:** 10.11.0 (works as expected in 10.10.0)
- **Vue version:** 3.5.42 (`vue-tsc` 3.3.11, `typescript` 6.0.3)
- **Node version:** 24.20.0
- **Operating System:** macOS 26.6.2

**Please show your full configuration:**

```js
// eslint.config.js
import tsParser from '@typescript-eslint/parser'
import pluginVue from 'eslint-plugin-vue'

export default [
  ...pluginVue.configs['flat/recommended'],
  {
    files: ['**/*.vue'],
    languageOptions: { parserOptions: { parser: tsParser } },
    rules: { 'vue/multi-word-component-names': 'off' },
  },
]
```

**What did you do?**

Since 10.11.0 (#3032), `vue/require-default-prop` requires a default on `defineModel`. The problem is that there is no way to satisfy it: the natural escape hatch, `default: undefined`, silences the rule but is not expressible in Vue's own types, and providing a real default contradicts Vue's documentation.

`Component.vue` — the reported code:

```vue
<script setup lang="ts">
const modelValue = defineModel<string | null>()
</script>

<template>
  <input :value="modelValue">
</template>
```

`Attempted.vue` — the attempted fix:

```vue
<script setup lang="ts">
const modelValue = defineModel<string | null>({ default: undefined })

function set(v: string) {
  modelValue.value = v
}
</script>

<template>
  <input :value="modelValue" @input="set('x')">
</template>
```

**What did you expect to happen?**

That `defineModel()` without a default is accepted, or at least that some type-safe, behaviour-preserving way to satisfy the rule exists.

The asymmetry with `defineProps` seems to be the crux. For a prop, "no default" just means the value is absent, and adding a default is normal. For `defineModel`, the absence of a default is what keeps parent and child in sync — Vue's docs warn against it explicitly:

> If you have a `default` value for `defineModel` prop and you don't provide any value for this prop from the parent component, it can cause a de-synchronization between parent and child components.

https://vuejs.org/guide/components/v-model.html#under-the-hood

This was raised on the PR after it merged: https://github.com/vuejs/eslint-plugin-vue/pull/3032#issuecomment-5572538993

**What actually happened?**

`npx eslint Component.vue` reports:

```
Component.vue
  2:20  warning  Prop 'modelValue' requires default value to be set  vue/require-default-prop

✖ 1 problem (0 errors, 1 warning)
```

Adding `default: undefined` silences ESLint, but `npx vue-tsc --noEmit` then fails. `DefineModelDefault<T>` resolves to `InferDefault<Data, T>`, i.e. `((props: Data) => T & {}) | (T extends NativeType ? T : never)`; `T & {}` excludes `undefined`, so neither `undefined` nor a `() => undefined` factory is assignable:

```
Attempted.vue(2,20): error TS2769: No overload matches this call.
  Overload 1 of 4, '(options: Omit<PropOptions<string | null, string | null>, "default"> & (DefineModelOptions<...> & ({ ...; } | { ...; }))): ModelRef<...>', gave the following error.
    Type 'undefined' is not assignable to type 'DefineModelDefault<string | null>'.
  Overload 2 of 4, '(options?: DefineModelRuntimeOptions<...> | undefined): ModelRef<...>', gave the following error.
    Object literal may only specify known properties, and 'default' does not exist in type 'DefineModelRuntimeOptions<string | null, string | null, string | null>'.
  Overload 3 of 4, '(name: string, options?: DefineModelRuntimeOptions<...> | undefined): ModelRef<...>', gave the following error.
    Argument of type '{ default: undefined; }' is not assignable to parameter of type 'string'.
Attempted.vue(5,3): error TS2322: Type 'string' is not assignable to type 'T'.
  'T' could be instantiated with an arbitrary type which could be unrelated to 'string'.
```

Note the second error: because the call falls off every overload, the returned ref degrades to a bare generic `T`, so every downstream assignment breaks too. In our codebase, adding `default: undefined` to the 8 `defineModel` declarations the rule flagged turned 0 type errors into 12 across 6 previously clean files.

That leaves no workable option:

| Attempt | Result |
| --- | --- |
| `default: undefined` | Silences ESLint, does not typecheck (above) |
| `default: () => undefined` | Same — `Type '() => undefined' is not assignable to type '(props: Data) => string'` |
| `required: true` | Typechecks, but changes runtime semantics: Vue warns whenever a parent omits an optional `v-model` |
| A real default | Typechecks, but is the de-sync footgun Vue's docs warn about |
| `eslint-disable` comment | Works, but one per `defineModel` call site |
| Turn the rule off | Works, but loses `defineProps` coverage; the rule has `schema: []`, so it cannot be disabled for `defineModel` alone |

**Suggested fix**

Drop `onDefineModelEnter` from `require-default-prop` and keep the `defineModel` support in `require-valid-default-prop` only, or gate it behind an opt-in option.

To be clear about scope: the `require-valid-default-prop` half of #3032 is good and caught a genuine bug for us (`defineModel<string[]>({ default: [] })` — a default array literal shared across instances). And that half is what the originating request, #2369, actually asked for: validate a default when one is deliberately given. Requiring one to exist is the part that has no valid answer.

**Repository to reproduce this issue**

Minimal reproduction, five files:

<details>
<summary><code>package.json</code></summary>

```json
{
  "name": "repro-require-default-prop-definemodel",
  "private": true,
  "type": "module",
  "devDependencies": {
    "@typescript-eslint/parser": "^8.70.0",
    "eslint": "^10.10.0",
    "eslint-plugin-vue": "^10.11.0",
    "typescript": "^6.0.3",
    "vue": "^3.5.42",
    "vue-tsc": "^3.3.11"
  }
}
```

</details>

<details>
<summary><code>tsconfig.json</code></summary>

```json
{
  "compilerOptions": {
    "strict": true,
    "noEmit": true,
    "module": "esnext",
    "moduleResolution": "bundler",
    "target": "es2022",
    "skipLibCheck": true
  },
  "include": ["*.vue"]
}
```

</details>

Plus `eslint.config.js`, `Component.vue` and `Attempted.vue` exactly as shown above. Then:

```sh
npm install
npx eslint Component.vue   # the warning with no valid fix
npx vue-tsc --noEmit       # why default: undefined is not that fix
```

Downgrading `eslint-plugin-vue` to 10.10.0 makes the ESLint warning disappear.

Happy to push this to a repository if a link is preferred.

**Comment by FloEdelmann** [comment]
Right, it indeed sounds like the `require-default-prop` change should be reverted and only the `require-valid-default-prop` change should remain.

Note that the comment in #3032 was also only raised _after_ merge and release.

**Comment by seanogdev** [comment]
PR open with the revert, good catch @hymair, apologies !
